integration_tests/fafclient.py

```diff
@@ -0,0 +1,155 @@
+import asyncio
+import subprocess
+from hashlib import sha256
+
+from server.protocol import QDataStreamProtocol
+from tests.integration_tests.conftest import read_until, read_until_command
+
+
+class FAFClient(object):
+    """docstring for FAFClient."""
+
+    def __init__(self, user_agent="faf-client", version="1.0.0-dev"):
+        self.proto = None
+        self.user_agent = user_agent
+        self.version = version
+        # Name of the faf-uid binary to call
+        # TODO: make this configurable via a command line arg?
+        self.faf_uid_path = "faf-uid"
+        self.player_id = None
+
+    def is_connected(self):
+        return self.proto and self.proto.is_connected()
+
+    def close(self):
+        if not self.is_connected():
+            return
+
+        self.proto.close()
+
+    async def connect(self, host, port):
+        self.proto = QDataStreamProtocol(
+            *(await asyncio.open_connection(host, port))
+        )
+
+    async def send_message(self, message):
+        """Send a message to the server"""
+        if not self.is_connected():
+            raise RuntimeError("Not connected to a server!")
+
+        await self.proto.send_message(message)
+
+    async def read_until(self, predicate, timeout=60):
+        return await asyncio.wait_for(
+            read_until(self.proto, predicate),
+            timeout=timeout
+        )
+
+    async def read_until_command(self, command, timeout=60):
+        return await read_until_command(self.proto, command, timeout=timeout)
+
+    # Commonly used functionality here
+    async def login(self, username, password):
+        """Perform the entire login sequence including fetching a session
+        and calling `faf-uid`"""
+        await self.proto.send_message({
+            "command": "ask_session",
+            "user_agent": self.user_agent,
+            "version": self.version
+        })
+        msg = await self.read_until_command("session")
+        session = msg["session"]
+        unique_id = self.get_unique_id(session)
+
+        pw_hash = sha256(password.encode("utf-8"))
+        await self.proto.send_message({
+            "command": "hello",
+            "version": self.version,
+            "user_agent": self.user_agent,
+            "login": username,
+            "password": pw_hash.hexdigest(),
+            "unique_id": unique_id
+        })
+        msg = await self.read_until_command("welcome")
+        self.player_id = msg["id"]
+        return msg
+
+    def get_unique_id(self, session):
+        try:
+            process = subprocess.run(
+                [self.faf_uid_path, str(session)],
+                stdout=subprocess.PIPE
+            )
+            return process.stdout.decode()
+        except OSError:
+            return "some_fake_uid"
```

Comment:
The uniqueness actually doesn't seem to matter since any results returned by the policy server are just ignored. The problem is if you send it data that isn't a valid uid then the policy server will return an http error which does cause the login to fail. So it doesn't matter at all what goes here, if this branch is reached then the login will fail.